### DISCUSSION TITLE
Draft state for snippets and quickly and accurately finding an object's usage

### DISCUSSION THREAD

**Discussion by kaedroho** [wagtail/wagtail#8496] [Ideas]
As part of an initiative to upgrade snippets in Wagtail (which would eventually lead to having Workflows on snippets!), we plan to add support for draft state to snippets.

We plan to introduce this feature as a mix-in that could be applied to snippet models that require draft state but you can still choose not to use this mix-in and snippets will behave the same as before. This mix-in will add the ``live``, ``has_unpublished_changes`` and ``live_revision`` fields that are currently only available for ``Page`` and enable UI in the admin for that snippet to allow managing draft snippets. The editor experience of draft snippets should be similar to pages.

We think that when a snippet is in draft state, it would not be selectable in any snippet choosers.

What should happen if a snippet that is currently selected on a live page is later unpublished?

### Object usage

We think it should be clear to the editor where any snippet is being used on a site, and warn/block them if they attempt to unpublish a snippet that is selected in any existing snippet chooser.

Wagtail has a [utility function](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/models.py#L27) that gets an objects usage and this is exposed in the admin for Images and Documents in their "usage" views. This method works by introspecting model definitions to find any ``ForeignKey`` that point at the given object's model, then performs a query on all those models to find objects that reference it.

This method does not check StreamFields or rich text fields, as these are more complex data types and we haven't got any method to unpack them in the database to find the references. Such a method could be very slow on large sites as the content of these fields can be quite big and we would be unable to use a database index.

#### Idea 1: Add a "references" model

This model would store all references from objects that are managed in Wagtail that could reference a snippet, image, or document. We would update the references at the point of saving an object that references other objects (probably using a ``post_save`` signal handler).

For example, the model might look something like:

```python
class Reference(models.Model):
    from_content_type = models.ForeignKey(ContentType, ...)
    from_object_id = models.CharField(...)
    to_content_type = models.ForeignKey(ContentType, ...)
    to_object_id = models.CharField(...)

    class Meta:
        indexed_together = [
            ('from_content_type', 'from_object_id'),
            ('to_content_type', 'to_object_id'),
        ]
```

Advantages:
 - This would be the fastest of all the methods I can think of. You can find all the back-references for any object by querying this model and filtering by the ``to_`` fields. This should also be faster than the current approach as we can find all references with a single model
 - This doesn't require changing the format of any fields

Disadvantages:
 - Slower insert performance, this could be an issue for bulk-imports or mass-copies
 - It will use some more disk space, but I expect this model would always be significantly smaller than the revision model

#### Idea 2: Change the format of StreamFields to allow extracting references efficiently

The current StreamField internal representation would be difficult to index as there are lots of different places where objects could be referenced.

To make indexing easier, we could change the format of StreamFields so that all references are listed in a top-level key. When you save a StreamField, all the external references would be repeated at this key, which could be indexed. For example:

```json
{
    "data": ["streamfield data here"],
    "refs": [
        "myapp.MyModel:2",
        "wagtailimages.Image:4",
    ]
```

Indexes could then be added to this field, for example in PostgreSQL the SQL would be: ``CREATE INDEX ON myapp_mypagemodel((body->>'refs'));``

Advantages:
 - Doesn't require a separate model to be kept in sync

Disadvantages:
 - Requires rewriting all existing StreamFields to update the format and can't be relied on for accuracy until then
 - The indexes need to be added to every StreamField manually
 - Doesn't help with RichTextFields or any custom field types that could also have references hidden in their content
 - You still need to query all ForeignKeys and now all StreamFields as well, so that query could be quite big for a complex site

#### Idea 3: Add a separate field for storing references

Similar to idea 2, but instead at the model level. This could probably be done with a mixin:

```python
# In Wagtail models

class ReferencesMixin(models.Model):
    references = models.JSONField()

    class Meta:
        abstract = True
        indexes = [
            Index('references'),  # or whatever the definition is!
         ]


@receiver(pre_save)
def update_references_field(instance, **kwargs):
    if not isinstance(instance, ReferencesMixin):
        return

    instance.references = extract_references(instance)
```

This mixin could be applied to the base ``Page`` model, and any snippet model that has outward references. We would probably need to add a system check to ensure this mixin gets applied on all models that needs it.

Advantages:
 - Doesn't require a separate model to be kept in sync
 - A single index for all references from a model (and this index could be shared for all pages)
 - This doesn't require changing the format of any fields

Disadvantages:
 - Still requires querying all models individually, but at least all pages would be combined, and we only need to check a single field for each model

---

Those are all the solutions I could think of for that problem!

What are your thoughts on this feature and the ideas listed above? Can you think of any other ideas on how to solve the object usage problem?

**Comment by gasman** [comment]
https://github.com/wagtail/wagtail/pull/4702 was a previous attempt at doing this without a supporting "references" table. It gets around the inefficiency of unpacking streamfield / rich text data in full by constructing a liberal regexp that fields containing the reference are guaranteed to match (but may include false positives), and then doing the full unpacking on the results.

I don't think there's anything fundamentally wrong with that approach - it just stalled because the original PR was written in a very dense and monolithic way and needs to be broken down into well-documented modules. Hopefully the move to JSONField should allow us to do that "first pass" query in a smarter way based on the JSON structure.

I'm a little wary of the reference-tracking method - it feels like it'll add a lot of moving parts to the low-level database handling. I guess it's fairly analogous to search indexing though, and can be approached in much the same way (i.e. post_save signals and a management command to rebuild the whole thing).

**Comment by jams2** [comment]
Hey @kaedroho, this has been in the back of my mind for a while, and coincidentally I started playing around with some parts of an implementation over the weekend! It has come up with clients a few times, for use cases like adding temporal metadata (e.g. licensing restrictions, content expiry dates) to objects, then flagging to content editors when an object meets some criteria, and where it is used. I have a few thoughts:

If performance of select queries is key it would be good to get a better idea of what support is like for indexing on JSON columns among the different database engines. Postgres seems reasonably flexible with the use of GINs on jsonb columns but from a brief search the situation with MySQL is not so promising.

I think the nature of object references in a Wagtail site is that they tend to be unique on a page level - I am more likely to include 5 distinct images on a page than I am to use the same image 5 times. With regards to storage space then, a solution will scale roughly linearly with the number of object references. Maybe there's a composite approach that allows for a kind of compression without sacrificing on performance, e.g.

```
reference table {
from_content_type :: id,
from_object_id :: id,
to_content_type :: id,
to_object_ids :: ids of all objects of to_content_type referenced on from_object, encoded and indexable
}
```

but that doesn't seem so great now that I've typed it out. It might make sense to try and optimise for the case of a single object that is referenced in multiple places.

One difference between using a references table and storing a JSON reference column will be deletion behaviour - with a reference table only that table will have to be queried and updated, whereas in the JSON reference column approach potentially many tables will have to be queried for references to the deleted object and updated.

If the trade-offs between the proposed approaches are significant enough, maybe it's worth implementing reference lookup functionality as a pluggable backend system, so implementers can store references (or not) how it suits them.

**Comment by cyface** [comment]
Perhaps I am missing something, but adding the 'References' model would also require some sort of batch / upgrade job to populate it for existing Pages, correct?  Since that was listed as a disadvantage for the Streamfield-rewrite approach, I thought it might be worth mentioning.

I also wonder if another possible approach would be an external index, much like we do with Postgres search, where both a crawler job and a post-save job could update an index, rather than an explicit data structure. 🤔
